### ISSUE TITLE
apply

### ISSUE THREAD

**Issue by technomancy** [meric/l2l#36]
I want to add an `apply` function, but I don't understand enough about the organization of the codebase to know where it should go.

**Comment by meric** [comment]
@technomancy Can you take a look at https://github.com/meric/l2l/tree/rewrite?

This branch will be replacing the master branch in entirety when it is ready.

The `apply` is not implemented yet, you're welcome to try.

Pure functions go in [lib](https://github.com/meric/l2l/tree/rewrite/l2l/lib)
Macros and special forms go in [lib](https://github.com/meric/l2l/tree/rewrite/l2l/ext)
Add the special forms to [https://github.com/meric/l2l/blob/rewrite/l2l/compiler.lua#L239](url), to make them loaded by default. Keep in mind, all extensions come with zero default extensions, and whatever you need must be imported manually, e.g. quote and quasiquote. There is currently a bug with quasiquote, not working inside lua statement nodes.

**Comment by meric** [comment]
For pure function libs that you think you be imported by default, the compiler can workout whether the library is being used in the source code using primitive static analysis, and write out the appropriate require statements.

See [initialize_dependencies](https://github.com/meric/l2l/blob/rewrite/l2l/compiler.lua#L121) function. Put the entry for your library in there.

The pseudo Lua "DSL" could be implemented better, but I think it's usable.

**Comment by meric** [comment]
In [fn.lisp](https://github.com/meric/l2l/blob/rewrite/l2l/ext/fn.lisp) that defines the `fn` special form which declares a function, that file implements how `fn` compiles into Lua. It doesn't compile to lua strings, it compiles to a lua abstract syntax tree.

You can see the definition of all the lua abstract syntax tree nodes here:
[Lua AST definition](https://github.com/meric/l2l/blob/rewrite/l2l/lua.lua#L56)

There are several types of nodes.

The Lua AST implementation uses the `ast` module in the `leftry` library:
[AST module](https://github.com/meric/leftry/blob/master/leftry/ast.lua)

It defines each type of AST node, how to construct them, etc.

**Comment by technomancy** [comment]
It looks like you have a broken symlink to `leftry`; maybe that should be a git subtree instead?

**Comment by meric** [comment]
Can you clone [leftry](https://github.com/meric/leftry) as a sibling repo to l2l, and re-create the symlink to it if necessary for now? I'm focusing on getting it into a turing-complete usable state before thinking about wrapping it up nicely...

**Comment by technomancy** [comment]
Here's what I get when I try to run the tests:

```
~/src/l2l $ lua l2l/test.lua 
/usr/bin/lua: ./l2l/compiler.lua:88: cannot not statize..nil
stack traceback:
	[C]: in function 'error'
	./l2l/compiler.lua:88: in function <./l2l/compiler.lua:64>
	(tail call): ?
	./l2l/compiler.lua:259: in function 'compile'
	./l2l/compiler.lua:296: in function 'compile_or_cached'
	./l2l/compiler.lua:214: in function 'build'
	./l2l/compiler.lua:225: in function 'import'
	./l2l/reader.lua:291: in function 'import_extension'
	./l2l/compiler.lua:252: in function 'compile'
	l2l/test.lua:101: in main chunk
	[C]: ?
```

**Comment by meric** [comment]
One sec, let me commit whats on my local.

**Comment by meric** [comment]
Try the latest commit.... I'm using Lua 5.2, and I confirm on my environment it works on LuaJIT (with lua 5.2. features enabled) also.

**Comment by technomancy** [comment]
Actually still getting the same error on 4ed860f with both LuaJIT 2.0.3 and Lua 5.1. Not sure whether my LuaJIT has the 5.2-compat flag though.

**Comment by meric** [comment]
I suspect you're using Lua 5.1 on both. It probably doesn't work on Lua5.1 because it doesn't have __ipairs for metatables implemented. It's possible to write some compatibility code so it works on Lua5.1 too, but it's not done yet.

**Comment by technomancy** [comment]
OK, yeah this is a major annoyance I have with the LuaJIT that ships with Love2d; I have had to monkeypatch it in my game in order for metamethods to be worth using: https://gitlab.com/technomancy/bussard/blob/master/metatable_monkey.lua

It makes me cringe so hard, but without that there is no way to make proxy tables. =\ At least it checks and only activates the monkeypatch if it's running under a lua that is missing the feature to begin with. I have tried to convince the Love2d devs to enable this flag, but they are not interested.

**Comment by meric** [comment]
Maybe it'll be possible to do some custom builds. Otherwise, have you tried submitting a pull request? https://bitbucket.org/rude/love 

I've been working on l2l because I wanted to use Lisp with Love2d. So when I'm done I'll get around to get Love2d to work well with Lua 5.2.

Then again, I've been working on l2l for the past few years, and this particular rewrite (including the parser combinator) since late last year... :)

**Comment by technomancy** [comment]
A custom love2d build would honestly probably just be more of a headache
at this point because I still want my game to work with stock love
that's installed via apt-get and possibly submit my game for inclusion
in Debian itself. So the monkeypatch isn't going away, sadly.

I've spoken with the love devs on IRC, and they seem to be wary of
potential compatibility issues, but on the other hand new versions of
love break compatibility with old versions all the time, so I don't
really understand that.

I had forgotten about the rewrite branch, but yesterday I ported one of
my kids' games to the old master branch of l2l: https://gitlab.com/technomancy/l2love/blob/master/examples/lospace/main.lisp

Anyway, I'm glad to hear that your goal is to make love games in lisp too!

Right now in the game I'm working on I want to make it possible to code
in lisp in-game, but I probably won't change the game engine to lisp. I
have basically ported Emacs to Lua, and it would be very cool to allow
configuration in lisp. (http://p.hagelb.org/bussard-split.png) But I
would love to use it for other games.

How far away is the rewrite from being usable? I am considering doing
ludum dare at the end of the month, and it would be cool to submit an
l2l-based entry.

**Comment by meric** [comment]
>  I've spoken with the love devs on IRC, and they seem to be wary of
potential compatibility issues

That's unfortunate. Maybe as Lua 5.1 get older more people will speak up and it'll be worth the effort to implement and test.

> https://gitlab.com/technomancy/l2love/blob/master/examples/lospace/main.lisp

Beautiful.

> How far away is the rewrite from being usable? 

There are some bugs to fix. It's technically usable now because all you have to do is write in Lua and put a backslash at the top of the file, so bugs aside, you can implement you need.

But I want to fill out the arithmetic operators, implement `if`, `let`, `apply`, etc.

What else do we need for a ludum dare entry? Let's make a list of lisp features you want, and we can prioritise those.

**Comment by technomancy** [comment]
> What else do we need for a ludum dare entry? Let's make a list of lisp
> features you want, and we can prioritise those.

https://github.com/meric/l2l/blob/master/l2l/compiler.lua#L121

If we had all the keywords here and fixed the 5.1 compatibility issues
that might be enough.

I can try porting over some toy examples at that point to see how
feasible it is.

There were littler things I found like not being able to let-bind
multiple return values, but those are minor luxuries:

```lisp
(let ((width height) (love.window.getMode))
  ...)
```

For an initial cut I am fine just pretending multiple values don't exist.

**Comment by meric** [comment]
Noted. I do encourage you to see if you can implement one or two. It would show me what issues others face when they do it.

**Comment by technomancy** [comment]
I wouldn't mind taking a swing at something.

When you say "pure functions go in lib" you're not talking about referential transparency, but rather the fact that they are functions rather than special forms, right?

I like the idea of doing as much as possible in .lisp functions and leaning heavily on the backslash stuff, but it makes syntax highlighting a huge pain.

Can you explain what the `hash` call at the end of arithmetic.lisp is about? Also I don't understand what is return from ext/fn.lisp; what do "expize" and "statize" mean? What is the comment at the top of arithmetic.lisp for?

**Comment by meric** [comment]
> rather the fact that they are functions rather than special forms

Correct. "Simple functions" might be better term for next time.

> I like the idea of doing as much as possible in .lisp functions and leaning heavily on the backslash stuff, but it makes syntax highlighting a huge pain.

At times the backslash stuff is still a bit iffy. It works nicely for ext/arithmetic.lisp, but I found some other cases where it doesn't work, yet.

At this time I suggest for extensions write it in Lua only, prepending a "\" at the top of the file. https://github.com/meric/l2l/blob/rewrite/l2l/ext/fn.lisp

> Can you explain what the `hash` call at the end of arithmetic.lisp is about?

A hash converts a symbol to Lua compatible form. E.g. "+" is not a valid Lua name, but is a valid Lisp name. The function is used to represent lisp names in Lua.

However, at that call, the hash is required only because when the reader does a lookup, [it uses `hash` to "decode" the macro name](https://github.com/meric/l2l/blob/rewrite/l2l/reader.lua#L269). Maybe it's possible to remove it both sides...Will investigate. Maybe all we need to do is convert symbol to its string name instead of hashing.

**Comment by meric** [comment]
There are two types of code in Lua. Expressions, and statements. Sometimes they're interchangeable, but often they're not. [Expize](https://github.com/meric/l2l/blob/rewrite/l2l/ext/fn.lisp#L67) should describe how the thing is compiled into a Lua expression. Statize should describe how the thing is compiled into a Lua statement.

**Comment by meric** [comment]
The comment at the top of arithmetic.lisp is me thinking whether to implement a `local` lisp keyword that declares local variables.

**Comment by technomancy** [comment]
Thanks for the info. One thing I noticed is there seems to be a lot less of a reliance of globals on the rewrite branch, which makes me glad.

> A hash converts a symbol to Lua compatible form. E.g. "+" is not a valid Lua name, but is a valid Lisp name. The function is used to represent lisp names in Lua.

Gotcha. I'm used to this being called "mangling"; the word "hash" has too many meanings. https://en.wikipedia.org/wiki/Name_mangling

> The comment at the top of arithmetic.lisp is me thinking whether to implement a local lisp keyword that declares local variables.

From my experience porting stuff to l2l, this would be really nice. Most lisps have a way to do top-level declarations that are not visible outside the module, but l2l just has let (which is fine for inside a function, but I don't like how it introduces a level of indentation for top-level stuff) and set.

**Comment by technomancy** [comment]
Implemented in #38.

**Comment by meric** [comment]
Great. I've also renamed hash to mangle.

**Comment by meric** [comment]
`local` is half done. https://github.com/meric/l2l/blob/rewrite/l2l/ext/local.lisp

I'm thinking what to do if someone does something like (print 1 2 (local a 1)). Currently it just errors.

**Comment by meric** [comment]
https://github.com/meric/leftry/commit/c360184e22acb0a6d3308c78e59a831af33974e5

Just a heads up Leftry got an update.

**Comment by technomancy** [comment]
If you think of `local` as being analogous to the first argument to the `let` macro, then I think you could make a good case for it being disallowed in an expression context.

On the other hand, one of the potential benefits of l2l is that you could pretend that there never was any statement/expression distinction in the first place. In this case it would be better for `local` to return nil in an expression context.

So I would say it comes down to what your philosophy of the relationship between l2l and Lua. If you lean more towards the "Lispiness" side, you will allow it, but if you lean more towards "Luaness", you won't.

**Comment by meric** [comment]
e8f93d2c90f3ac88c18a70e34dc8495f1f15c48b

Lisp has Lispiness. Lua has Luaness, and yet they are one.

[(print (local c d 1))](https://github.com/meric/l2l/blob/rewrite/l2l/test.lua#L9)

compiles into 

```
local c,d = 1
print(c,d)
```

I have made an update to leftry also.

**Comment by technomancy** [comment]
Very nice.

Would you like me to submit a PR changing leftry into a git subtree so it is kept in sync automatically?

**Comment by meric** [comment]
Thanks. Yes please!

**Comment by technomancy** [comment]
Hm; this is a little bit awkward because if `leftry` is inside the `l2l` dir then we need to require `leftry.leftry` and `leftry.leftry.elements...`. Maybe instead of having a `leftry.lua` file and a `leftry` directory inside the `leftry` directory, there could just be `init.lua` with an `elements` directory inside that?

**Comment by meric** [comment]
When leftry is implemented I'll remove the leftry reference inside l2l. Then it will be up to the user to either install it using `luarocks install --server=http://luarocks.org/dev leftry` or to clone the leftry repository at the top level of their project so they can do `require('leftry')`.

**Comment by meric** [comment]
Perhaps we could do with a Makefile.?

**Comment by technomancy** [comment]
I guess it depends on how long it will be before you spin it out into luarocks. But I don't think that it's a good fit for git subtree at this point due to the organization of the project.

In any case, a makefile is a good idea anyway just so it's clear how to run the tests.

You're probably also going to want to get it so the tests can actually report when they fail. I have had great luck with lunatest in my game: https://github.com/silentbicycle/lunatest I would also strongly recommend adding a makefile target for luacheck, since I have found this to catch a lot of potential bugs in my own project. I wouldn't mind taking a shot at cleaning up the luacheck warnings if you are OK with the conventions it recommends and would be interested in keeping the codebase warning-free.

**Comment by meric** [comment]
Ok. I wouldn't mind at all. :)

**Comment by technomancy** [comment]
At the top of `compiler.lua` there are a bunch of locals like `local lua_functioncall = lua.lua_functioncall`. Under PUC Lua this can give you a slight performance boost, but under LuaJIT it actually can prevent certain optimizations, and my understanding is that it's usually unhelpful. Do you mind if I remove these and just replace them with calls to the original functions?

I'm already seeing a few places where the code refers to undefined variables; I am not really sure what the correct fix is, so I'll just mark them with TODOs.

**Comment by meric** [comment]
Ok. I don't mind, it's mostly aliasing so I didn't have to type "lua.lua" which seemed repetitive.

**Comment by meric** [comment]
@technomancy 

Ludum Dare is next week.

Most of the [keywords](/meric/l2l/tree/rewrite/l2l/ext) in old l2l have been implemented, except the looping special forms. I don't think they will be ready by next week, I got some neat ideas that will take some time to experiment - right now the only way to loop is to escape into Lua.

If you choose to use l2l for Ludum Dare, with a week to go, my opinion is...maybe making sure the current l2l is bug free enough, and actually ready to be used, is more important than implementing looping constructs that are likely to be buggy this time next week.

In conclusion, I think if l2l rewrite branch is either ready by now, minus a few tweaks and bug fixes, or it isn't and I want to give you time to look for alternatives.
